src/vm/inlinetracking.cpp

```diff
@@ -418,12 +424,181 @@ void PersistentInlineTrackingMap::Save(DataImage *image, InlineTrackingMap* runt
+    // and create an InlineeRecord with correct offset
+    InlineeRecordR2R record(RidFromToken(inlinee.m_methodDef));
+    DWORD inlinersStreamSize;
+    const BYTE *inlinersStreamPtr = (const BYTE *)inlinersStream.GetBlob(&inlinersStreamSize);
+    record.m_offset = inlinersBuffer->GetSize();
+    inlinersBuffer->Insert(inlinersBuffer->End(), SBuffer(SBuffer::Immutable, inlinersStreamPtr, inlinersStreamSize));
+
+    inlineeIndex->Append(record);
+}
+
+void PersistentInlineTrackingMapR2R::Save(SBuffer* pSaveTarget, InlineTrackingMap* runtimeMap)
 {
     STANDARD_VM_CONTRACT;
-    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMap, m_module));
-    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMap, m_inlineeIndex));
-    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMap, m_inlinersBuffer));
+    _ASSERTE(pSaveTarget != NULL);
+    _ASSERTE(runtimeMap != NULL);
+
+    SArray<InlineeRecordR2R> inlineeIndex;
+    SBuffer inlinersBuffer;
+
+    // Sort records from runtimeMap, because we need to make sure 
+    // we save everything in deterministic order. Hashtable iteration is not deterministic.
+    COUNT_T runtimeMapCount = runtimeMap->GetCount();
+    InlineTrackingEntry **inlinees = new InlineTrackingEntry *[runtimeMapCount];
+    NewArrayHolder<InlineTrackingEntry *>inlineesHolder(inlinees);
+    int index = 0;
+    for (auto iter = runtimeMap->Begin(), end = runtimeMap->End(); iter != end; ++iter)
+    {
+        inlinees[index++] = const_cast<InlineTrackingEntry *>(&*iter);
+    }
+    util::sort(inlinees, inlinees + runtimeMapCount, compare_entry);
+
+
+    // Iterate throught each inlinee record from the InlineTrackingMap
+    // and write corresponding records into inlineeIndex and inlinersBuffer
+    for (COUNT_T i = 0; i < runtimeMapCount; i++)
+    {
+        ProcessInlineTrackingEntry(&inlinersBuffer, &inlineeIndex, inlinees[i]);
+    }
+
+    InliningHeader header;
+    header.MajorVersion = 1;
+    header.MinorVersion = 0;
+    header.SizeOfInlineeIndex = inlineeIndex.GetCount() * sizeof(InlineeRecordR2R);
+    
+    pSaveTarget->Insert(pSaveTarget->End(), SBuffer(SBuffer::Immutable, (const BYTE*) &header, sizeof(header)));
+    DWORD unused = 0;
+    pSaveTarget->Insert(pSaveTarget->End(), SBuffer(SBuffer::Immutable, (const BYTE*) inlineeIndex.GetElements(), header.SizeOfInlineeIndex));
+    pSaveTarget->Insert(pSaveTarget->End(), SBuffer(SBuffer::Immutable, (const BYTE*) inlinersBuffer, inlinersBuffer.GetSize()));
+
+    LOG((LF_ZAP, LL_INFO100000,
+        "PersistentInlineTrackingMap saved. InlineeIndexSize: %d bytes, InlinersBufferSize: %d bytes\n",
+        header.SizeOfInlineeIndex, inlinersBuffer.GetSize()));
 }
+
 #endif //FEATURE_NATIVE_IMAGE_GENERATION
+
+BOOL PersistentInlineTrackingMapR2R::TryLoad(Module* pModule, const BYTE* pBuffer, DWORD cbBuffer)
+{
+    InliningHeader* pHeader = (InliningHeader*)pBuffer;
+    if(pHeader->MajorVersion != 1)
```

Comment:
>MajorVersion [](start = 16, length = 12)

Have you considered using the R2R major and minor versions instead of declaring new ones? (also seems like MinorVersion is not used)